File: services/ai-gateway/app/schemas/dialogue.py
"""口语陪练 Schema"""
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class DialogueStartRequest(BaseModel):
    user_id: str
    scenario: Literal[
        "interview",      # 面试
        "travel",         # 旅行
        "business",       # 商务
        "academic",       # 学术
        "daily",          # 日常
    ] = "interview"
    level: Literal["A2", "B1", "B2", "C1", "C2"] = "B2"
    language: Literal["zh-CN", "en-US"] = "en-US"


class DialogueMessage(BaseModel):
    role: Literal["assistant", "user"]
    content: str
    pronunciation_score: float | None = None  # 0-100（仅 user 消息）
    corrections: list[str] = []  # 仅 user 消息，含纠错


class DialogueStartResponse(BaseModel):
    session_id: str
    scenario: str
    level: str
    opening: str  # AI 开场白


class DialogueTurnRequest(BaseModel):
    session_id: str
    user_text: str  # 用户说的文字（已 ASR 转换）
    audio_url: str | None = None


class DialogueTurnResponse(BaseModel):
    assistant_text: str
    corrections: list[str] = []
    vocabulary_used: list[str] = []  # 用户使用的新词
    next_question: str  # AI 给出的追问


class DialogueScoreRequest(BaseModel):
    session_id: str
    transcript: list[DialogueMessage]


class DialogueScoreResponse(BaseModel):
    overall_score: float = Field(..., ge=0, le=100)
    pronunciation: float
    grammar: float
    vocabulary: float
    fluency: float
    highlights: list[str] = []
    improvements: list[str] = []

File: apps/web/src/services/dialogue.ts
/**
 * 口语对话陪练 API
 */
import { config, fetchWithFallback } from '../config'

export type Scenario = 'interview' | 'travel' | 'business' | 'academic' | 'daily'
export type Level = 'A2' | 'B1' | 'B2' | 'C1' | 'C2'

export interface DialogueStartResult {
  session_id: string
  scenario: string
  level: string
  opening: string
}

export interface DialogueTurnResult {
  assistant_text: string
  corrections: string[]
  vocabulary_used: string[]
  next_question: string
}

export interface DialogueMessage {
  role: 'assistant' | 'user'
  content: string
}

export interface DialogueScoreResult {
  overall_score: number
  pronunciation: number
  grammar: number
  vocabulary: number
  fluency: number
  highlights: string[]
  improvements: string[]
}

const BASE = '/api/ai/dialogue'

export async function startDialogue(
  scenario: Scenario,
  level: Level,
): Promise<DialogueStartResult> {
  const path = `${BASE}/start`
  const directUrl = config.aiGateway ? `${config.aiGateway}${path}` : path
  const resp = await fetchWithFallback(path, directUrl, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ user_id: 'web', scenario, level, language: 'en-US' }),
  })
  if (!resp.ok) throw new Error(`开始对话失败 ${resp.status}`)
  return resp.json()
}

export async function turnDialogue(
  sessionId: string,
  userText: string,
): Promise<DialogueTurnResult> {
  const path = `${BASE}/turn`
  const directUrl = config.aiGateway ? `${config.aiGateway}${path}` : path
  const resp = await fetchWithFallback(path, directUrl, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ session_id: sessionId, user_text: userText }),
  })
  if (!resp.ok) throw new Error(`对话失败 ${resp.status}`)
  return resp.json()
}

export async function scoreDialogue(
  sessionId: string,
  transcript: DialogueMessage[],
): Promise<DialogueScoreResult> {
  const path = `${BASE}/score`
  const directUrl = config.aiGateway ? `${config.aiGateway}${path}` : path
  const resp = await fetchWithFallback(path, directUrl, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ session_id: sessionId, transcript }),
  })
  if (!resp.ok) throw new Error(`评分失败 ${resp.status}`)
  return resp.json()
}

export const SCENARIOS: { id: Scenario; label: string; emoji: string; desc: string }[] = [
  { id: 'interview', label: '面试', emoji: '💼', desc: '模拟求职面试' },
  { id: 'travel', label: '旅行', emoji: '✈️', desc: '旅行出行场景' },
  { id: 'business', label: '商务', emoji: '📈', desc: '商务会议沟通' },
  { id: 'academic', label: '学术', emoji: '🎓', desc: '学术讨论交流' },
  { id: 'daily', label: '日常', emoji: '☕', desc: '日常生活闲聊' },
]

export const LEVELS: Level[] = ['A2', 'B1', 'B2', 'C1', 'C2']
